Added edge gestures, a sets of gestures that only work at the top, left and right edge of the screen (#20551)
nvaccess/nvda: user_docs/en/changes.md

* Added sequential two-flick touch gestures that combine two flicks performed in quick succession into a single gesture, increasing the number of touch gestures that can be bound to scripts. (#19938, @kefaslungu)
  * Twelve combinations are recognised: opposite-direction pairs (e.g. flick right then flick left) and perpendicular L-shaped pairs (e.g. flick right then flick up).
  * The two flicks can be performed either by lifting the finger between strokes or as a single continuous swipe with a sharp change in direction.
* Added edge gesture support for touch screens, allowing gestures that begin within 15 mm of the left, right, or top screen edge to be bound independently from the same gesture performed in the centre of the screen. (#19938, @kefaslungu)
  * Edge gestures are disabled by default and can be enabled in the Touch Interaction settings panel.

>   * The bottom edge is intentionally excluded as Windows and the taskbar intercept touches there.

Reviewer: what happens if you change your taskbar position?

>   * The bottom edge is not supported; Windows intercepts swipe-from-bottom gestures, and the taskbar (which may be on any edge) will also intercept touches on whichever edge it occupies.

Reviewer: can you also update the user guide to reflect these changes to the bottom edge support you just pushed?

>   * The bottom edge is not supported; Windows intercepts swipe-from-bottom gestures, and the taskbar (which may be on any edge) will also intercept touches on whichever edge it occupies.

Reviewer: if the taskbar will intercept touches from the left edge if its there, does that mean NVDA and the taskbar will conflict? what will happen? does this mean if you move the taskbar you can only use 2 edges?

Author: On Windows 11, the taskbar is locked to the bottom and cannot be moved, so this is not a concern there. On Windows 10, if a user moves their taskbar to another edge, Windows will intercept touches on that edge before NVDA receives them — there is no conflict or crash, that edge simply becomes unavailable for NVDA gestures. We could consider documenting this caveat more explicitly in the user guide if that would help.

Reviewer: >On Windows 11, the taskbar is locked to the bottom and cannot be moved, so this is not a concern there.

That's not true, I'm on Windows 11 and can change it through Settings, Personalisation, Taskbar, taskbar behaviours, task bar position.

>On Windows 10, if a user moves their taskbar to another edge, Windows will intercept touches on that edge before NVDA receives them — there is no conflict or crash, that edge simply becomes unavailable for NVDA gestures.

On Windows 10, is Windows intercepting the bottom edge and the other edge, or just the edge with the taskbar?

Reviewer: I'm asking this, because if the taskbar position affects windows interception, we need to make this smarter. we can get the taskbar position with [the windows API](https://learn.microsoft.com/en-us/windows/win32/api/shellapi/nf-shellapi-shappbarmessage). We should probably allow all gestures, with a warning about the taskbar if we detect a clash. I'm not sure how we should handle clashes.

Author: I've just checked on Windows 11 25H2 — the option under Settings → Personalisation → Taskbar → Taskbar behaviours labelled "Taskbar position" only controls icon alignment (Left or Center), not which screen edge the taskbar sits on. The taskbar remains locked to the bottom on Windows 11 and cannot be moved to another edge through standard settings. All I see is taskbar alignment.

Reviewer: Can you answer the question on how we handle Windows 10 case? This is a `26H1` (preview) feature, so it will become an issue on Windows 11 soon.

Author: We temporarily implemented bottom edge gesture support to test the real behaviour. Here are today's findings:

1. **Bottom edge is intercepted by Windows Shell at the system level**, completely independent of the taskbar. Swiping from the bottom-left opened the Start menu, and from the bottom-right opened the Action Center. NVDA received no bottom edge touches at all.

2. **This is not a taskbar issue.** The interception happens even when the taskbar is not involved — it is a Windows Shell gesture feature.

3. **Screen rotation confirms it.** Testing was done on a convertible device. When the device was folded so that the physical top became the physical bottom, the previously working top edge stopped working, and the previously non-working bottom edge started working. This confirms Windows intercepts the physical bottom edge regardless of screen orientation or NVDA coordinate system.

4. **Other edges work fine.** Left, right, and top edges are not intercepted by Windows Shell and NVDA receives gestures from them correctly.

Since you have access to 26H1 preview where the taskbar can be repositioned, could you test moving the taskbar to another edge and confirm whether NVDA still receives swipe gestures from that edge? This will help us determine whether taskbar position affects gesture reception on non-bottom edges.

Reviewer: In Windows 11, moving the taskbar also changes the interception. i.e. if moving the taskbar to the top of the screen, swipes down from the top edge open the start menu.
This means I think we need to support all 4 edge gestures and we need to be aware of where the taskbar is

Reviewer: I think for Windows older than 26H1, we disable the bottom gesture.
For Windows 26H1 and newer, we allow mapping all gestures, and try to detect/avoid collisions where possible when setting or attempting gestures

Reviewer 2: We don't do this for any other input gestures where they may conflict with system gestures, do we?

Reviewer: That's true @SaschaCowley - do you think we need to do anything here then? I think we should support all 4 gestures, and if you run into conflicts well don't bind it? You will discover how the taskbar works pretty quickly.

Reviewer 3: So, how about hiding the taskbar?

Author: Agreed with @SaschaCowley and @seanbudd — we now support all four edges without any special detection logic. This is consistent with how NVDA handles other gesture conflicts with system shortcuts: if an edge is intercepted by Windows or the taskbar, NVDA simply never receives the touch and the user will discover it naturally.

Good question @hwf1324 — unfortunately hiding the taskbar (auto-hide) does not free the edge, as Windows still intercepts the swipe to reveal it. Fully removing the taskbar would work, but that is not something most users would want to do just to use edge gestures.

>   * All four edges are supported. Note that Windows or the taskbar may intercept gestures on some edges (e.g. the bottom edge opens the Start menu or Action Center); if an edge is intercepted, NVDA will not receive those gestures.

Reviewer: ```suggestion
  * All four edges are supported.
  Note that the Windows taskbar may override gestures on an edge.
  Gestures from the taskbar edge open the Start menu or Action Center, NVDA will not receive them.
```
